src/kernels/gemm.cu

```diff
@@ -0,0 +1,556 @@
+    using SharedB =
+        SharedTile<InType, tl::ColMajor<kTK, kTN>, kSwizzled, kSharedAccess>;
+    // Access a single register tile for operand B
+    using SIteratorB = STileIterator<SharedB, TileShape<kRK, kTN>>;
+
+    static_assert(GIteratorA::sc1 == GIteratorB::sc0,
+                  "mismatched K dimension!");
+    static_assert(SIteratorA::sc1 == SIteratorB::sc0,
+                  "mismatched K dimension!");
+
+    // Register tile for a single thread of operand A
+    static constexpr int kBKs = kRK / BaseShape::kRows;
+    static constexpr int kBNs = kTN / kWarpPerCol / BaseShape::kCols;
+    using RegB = RegTile<BaseTileColMajor<InType>, tl::ColMajor<kBKs, kBNs>>;
+
+    using G2SLoaderB = GlobalToSharedLoader<SharedB, WarpLayout>;
+    using S2RLoaderB =
+        SharedToRegLoader<RegB, WarpLayout, WarpReuse::kColReuseCont>;
+
+    // Global Tile for output C
+    using GlobalC = GlobalTile<AccType, tl::RowMajor<kTM, kTN, kN>>;
+    // Shared Tile for output C
+    using SharedC = SharedTile<AccType, tl::RowMajor<kTM, kTN>, kSwizzled>;
+
+    // Register Tile for output C
+    static constexpr int kCMs = kTM / kWarpPerRow / BaseShape::kRows;
+    static constexpr int kCNs = kTN / kWarpPerCol / BaseShape::kCols;
+    using RegC = RegTile<BaseTileRowMajor<AccType>, tl::RowMajor<kCMs, kCNs>>;
+
+    using R2GStorerC = RegToGlobalStorer<GlobalC, RegC, WarpLayout>;
+    using R2SStorerC = RegToSharedStorer<RegC, WarpLayout>;
+    using S2GStorerC = SharedToGlobalStorer<SharedC, WarpLayout>;
+
+    using PipelineG2SA = Pipeline<InType, GlobalA, SharedA, GIteratorA,
+                                  G2SLoaderA, kNumStages, GIteratorA::sc1>;
+    using PipelineG2SB = Pipeline<InType, GlobalB, SharedB, GIteratorB,
+                                  G2SLoaderB, kNumStages, GIteratorA::sc1>;
+
+    using PipelineS2RA = Pipeline<InType, SharedA, RegA, SIteratorA, S2RLoaderA,
+                                  kNumStages - 1, SIteratorA::sc1>;
+    using PipelineS2RB = Pipeline<InType, SharedB, RegB, SIteratorB, S2RLoaderB,
+                                  kNumStages - 1, SIteratorA::sc1>;
+};
+
+template <typename InType, typename AccType,                  //
+          const int kM, const int kN, const int kK,           //
+          const int kTM, const int kTN, const int kTK,        //
+          typename GIteratorA, typename SIteratorA,           //
+          typename SharedA, typename RegA,                    //
+          typename G2SLoaderA, typename S2RLoaderA,           //
+          typename GIteratorB, typename SIteratorB,           //
+          typename SharedB, typename RegB,                    //
+          typename G2SLoaderB, typename S2RLoaderB,           //
+          typename GlobalC, typename SharedC, typename RegC,  //
+          typename R2SStorerC, typename S2GStorerC>
+__global__ void ke_gemm_stage1(const InType* dA, const InType* dB,
+                               AccType* dC) {
```

Comment:
`stage1` represents no pipelining. `stage2` incorporates pipelining of some loops using global-to-shared asynchronous copy instructions. `stage3` involves additional pipelining by pre-fetching data from shared memory to registers for some loops. I recommend using the term 'level' instead of 'stage' to avoid confusion with the conventional `num_stages` for GEMM kernels.